jeison-vector-of, jeison-hashtab-of type constraints and other enhancements

bscottm opened the pull request:
- Support ":type (jeison-vector-of type)" for vectors of things.
- Support ":type (jeison-hashtab-of type) for slots that contain hash tables.
- Support ":type (or null type)" and ":type (or type null)" so that you can initialize a slot to nil, but get something else when the JSON is read/interpreted (like hash tables or detect optional attributes in the JSON stream.
- Support using both native JSON and json.el.
- Can also pass a buffer to `jeison-read`, in addition to strings and alists. 

Apologies for reformatting README.md. Needed to make it easier for me to read.

There are two test cases for `jeison-hashtab-of` in `jeison-test.el`.

NOTE: The type constraints are prefixed to avoid future name collisions or conflicts. "vector-of" and "hashtab-of" would be good extensions to EIEIO.

bscottm committed:
1.1.0 enhancements

Consolidated commit with:

- Support ":type (jeison-vector-of type)" for vectors of things.
- Support ":type (jeison-hashtab-of key-type value-type) for slots that
  contain hash tables.
- Support ":type (or null type)" and ":type (or type null)" so that
  you can initialize a slot to nil, but get something else when the
  JSON is read/interpreted.
- Support using both native JSON and json.el.

NOTE: The type constraints are prefixed to avoid future name collisions
or conflicts. "vector-of" and "hashtab-of" would be good extensions to
EIEIO.

SavchenkoValeriy commented on README.md:
````diff
@@ -295,16 +392,96 @@ Here is an example:
+			}
+		}
+	}
+```
````
In all previous examples I used spaces for indentation, so maybe here we can also do that?

SavchenkoValeriy replied:
Actually, maybe we can step away from a real world example and "dumb it down" similarly to how it is done in the rest of README? It seems a bit too specific for CMake.

SavchenkoValeriy commented on jeison.el:
```diff
@@ -121,23 +186,47 @@ your needs, please consider declaring a separate class."
         (jeison--class-slots class-name)))
 
-(defun jeison-read (type alist-or-json &optional path)
+(defun jeison-read (type inp-thing &optional path)
```
Let's think of a better name than `inp-thing`, this function is user-facing and should be more descriptive.

SavchenkoValeriy commented on jeison.el:
```diff
@@ -121,23 +186,47 @@ your needs, please consider declaring a separate class."
+                                        (and (stringp inp-thing)
+                                             #'jeison--read-json-string)
+                                        #'identity)
+                                    inp-thing)
```
Let's use `cond` clause for this: https://www.gnu.org/software/emacs/manual/html_node/elisp/Conditionals.html#index-cond it will make it easier to read. And I prefer it back in the `let`, overcrowded lisp expressions can be very painful to read.

SavchenkoValeriy commented on jeison.el:
```diff
@@ -121,23 +186,47 @@ your needs, please consider declaring a separate class."
+        (satisfies ,(lambda (vec)
+                      (cl-every (lambda (elem) (cl-typep elem elem-type))
+                                vec)))))
+(cl-deftype jeison-hashtab-of (elem-type)
```
Since the predicate is `hash-table-p`, let's match it in `jeison` API: `jeison-hash-table-of`

SavchenkoValeriy commented on test/jeison-test.el:
```diff
@@ -218,4 +218,59 @@
+  (jeison-defclass jeison:jeison-class nil
+    ((reply :initarg :reply :initform nil :type (or null (jeison-hashtab-of (jeison-hashtab-of jeison:jeison-thing))))))
+  (let ((answer (jeison-read jeison:jeison-class "{
+	\"reply\" :
```
This is a tab :-)

SavchenkoValeriy commented on test/jeison-test.el:
```diff
@@ -218,4 +218,59 @@
+	    }
+        }
+}")))
+    (message "reply %s" (pp-to-string (oref answer reply)))
```
Seems like some debug logging left in here

SavchenkoValeriy commented on jeison.el:
```diff
@@ -121,23 +186,47 @@ your needs, please consider declaring a separate class."
+                        (maphash (lambda (_key value) (and snickers (cl-typep value elem-type)))
+                                 htab)
+                        ;; ... because snickers really satisfies (sm). :-)
+                        snickers)))))
```
I'm having trouble understanding this code here. Is `snickers` always `t`? Do we always return `t` here?

bscottm replied:
`maphash` always returns `nil`, which causes the type constraint to always fail. `snickers` is there to accumulate a return value, which will be `t` in the nominal case.  And, yes, it's a bit tongue in cheek.

SavchenkoValeriy replied:
I mean, I get that `maphash` always returns `nil` (it's in the comment :-) ). I don't see where we mutate `snickers`, it looks like `snickers` is `t`, inside of the `map` we do `(and snickers REAL-COND)`, which is identical to `REAL-COND`, and then throw away results of this expression. And in the end we return `snickers`, which is (and has always been) `t`. Am I missing something here?

SavchenkoValeriy replied:
@bscottm what about this conversation?

bscottm replied:
Ah, there's a missing `setq`.

SavchenkoValeriy replied:
Feels like we are missing a test case then :-)

SavchenkoValeriy commented on jeison.el:
```diff
@@ -148,30 +237,50 @@ JSON is an `alist' representing a JSON object where we want information
-             ;; and parse element-type (retrieved from the type)
-             ;; from each element
-             (mapcar (lambda (element)
-                       (jeison--read-internal element-type element)) json))
```
I'm not a big fan of testing the same thing twice. Maybe it would be better to extract this whole piece into `jeison--read-sequence`, test independently for list and vector, and for vector case wrap the call to that new function with `vector`?

bscottm replied:
The `mapcar` is your original code. :-)

The `let` just assigns the resulting list to `read-data`, which either gets returned unmolested in the `list-of` case or converted to `vector` in the `jeison-vector-of` case. Perhaps it could be shifted; I wanted to reduce the amount of duplicated code.

SavchenkoValeriy replied:
This comment is obviously misplaced, but my point still stands. In the `cond` condition, you check for `(or IS-LIST IS-VECTOR)` and then at the end of the case, you check which one of them was it. I do understand that it is for reducing duplication and I suggest to extract a common part into a separate function `jeison--read-sequence` (or `jeison--read-array` to be closer to JSON terminology). Then condition will look very simple:
```
(`(list-of ,element-type) (jeison--read-sequence ARGS))
(`(jeison-vector-of ,element-type) (vector (jeison--read-sequence ARGS))
```
I believe that this way is cleaner.

SavchenkoValeriy replied:
@bscottm and this?

bscottm replied:
???

SavchenkoValeriy commented on test/jeison-test.el:
```diff
@@ -218,4 +218,59 @@
                                   }"
                           '(a (jeison:filter-candidates b) number))))))
 
+(ert-deftest jeison:check-hashtab-of-type ()
```
Can we please have separate tests for optional fields?

bscottm replied:
I dunno. Can we? :-)

SavchenkoValeriy commented on jeison.el:
```diff
@@ -226,10 +335,13 @@ in the target class' constructor:
+  (let ((slot-value (jeison--read-internal
+                     (oref slot type) json (oref slot path))))
+    (if (null slot-value)
+        nil
```
I don't see where we check that the slot has `initform` here. What happens if the user DOES expect to have some value, will we silently put `nil` there? Maybe we should have `optional-type` for slots that can be nullable?

bscottm replied:
So, that's the 3 year old pull request that I incorporated. The intent is to ensure that the `:initform` value is left unmolested, although, now that you point it out, it'll be overwritten by `nil`. Presumably, the EIEIO object has already been constructed, so grabbing the slot's existing value might be the better solution.

SavchenkoValeriy commented on README.md:
````diff
@@ -295,16 +392,96 @@ Here is an example:
+  ((reply :initarg :reply
+          :initform nil
+          :type (or null (jeison-hashtab-of (jeison-hashtab-of cmake-fapi-reply-client-data))))))
+```
````
In this example it is simply not clear where to look.

bscottm replied:
Replaced with an expanded "Skills" example.

SavchenkoValeriy commented on README.md:
```diff
@@ -295,16 +392,96 @@ Here is an example:
+          :type (jeison-hashtab-of (jeison-hashtab-of cmake-fapi-reply-client-data)))))
+
+(cl-defmethod initialize-instance :after ((self cmake-fapi-reply) &rest _args)
+  (setf (oref self reply) (make-hash-table :test 'equal)))
```
Why does the user do all that, can we do it automagically if we see the `hash-table` type requirement?

bscottm replied:
Because `:initform` is quotes its arguments.

bscottm replied:
OBE. Example reworked.

SavchenkoValeriy commented:
Thanks for working on it! I have a few questions/suggestions about API convenience and code readability. And for the future, it would've been easier to review if all new features had been separated into separate PRs. Cheers!

bscottm commented:
Each commit corresponds to a comment, to help simplify the review. The code comments should be resolved. I'll work the README.md comments in the near future. And the `(or null type)` ERT test.

bscottm commented:
Separate PRs: Since the last pull request was close to 3 years old, I figured that you'd gone off to do more interesting things, so I was willing to take over maintenance.

SavchenkoValeriy commented on jeison.el:
```diff
@@ -198,12 +198,12 @@ PATH is a `list' of keys we should consequently find in JSON and
-                                             #'jeison--read-json-string)
-                                        #'identity)
-                                    inp-source)
+                         (funcall (cond
```
I also meant to use cond, to get rid of `funcall`, it's really unnecessary here

bscottm replied:
No, `funcall` is needed. The first argument of an eval-ed list must be a function. `cond` is not a function, unless you meant rewriting as:

```emacs-lisp
(cond ((bufferp inp-source) (jeison--read-buffer inp-source))
       ((stringp inp-source) (jeison--read-string inp-source))
       (t inp-source))
```

Which is a bit repetitive. `funcall` makes it more compact and idiomatic Lisp (FWIW, I've been coding in CL since ~1985.)

Also, in the Emacs source itself, there are a few places where this idiom is used, e.g., ido.el, around line 2395. So, it's not unprecedented.

SavchenkoValeriy replied:
Yeah, I meant rewriting it that way simply because it's one less level of indirection. I like simple dumb code. Because of this I don't like using old code or other code as a benchmark of what is considered normal in terms of style. In many cases, old code tends to be too smart and too convoluted for its own good.

bscottm replied:
You're taking my Lisp-foo magic away!

SavchenkoValeriy commented on jeison.el:
```diff
@@ -186,24 +186,24 @@ your needs, please consider declaring a separate class."
 
-(defun jeison-read (type inp-thing &optional path)
-  "Read TYPE from ALIST-OR-JSON by the given PATH.
+(defun jeison-read (type inp-source &optional path)
```
I feel bad for harassing you on that, but it is still not there. What is `inp`? Is it `input`? Then call it `input`, it's just two extra characters. But going further, what **is** `input-source`, it feels a bit tautological. Maybe we can call it `input-json` and then it kind of obvious.

bscottm replied:
But if you call it `input-json`, you lose the alist part. :-)

Sure, if you want `input-source`, it can be `input-source`. `inp` has always been an abbreviation for input. And if you look around enough CL source, `thing` tends to be used when describing a polymorphic entity.

SavchenkoValeriy replied:
`alist` is yet another representation of JSON, native Emacs Lisp JSON :-) So, I don't see this as a problem.

SavchenkoValeriy commented:
> Each commit corresponds to a comment, to help simplify the review. The code comments should be resolved. I'll work the README.md comments in the near future. And the `(or null type)` ERT test.

There are a couple of code comments that are still unresolved

SavchenkoValeriy commented:
> Separate PRs: Since the last pull request was close to 3 years old, I figured that you'd gone off to do more interesting things, so I was willing to take over maintenance.

Yeah, no worries on that part. I understand

bscottm commented:
Ok, I think everything's been addressed.

bscottm commented:
Just rebased off your master, hence the forced push.

bscottm commented:
@SavchenkoValeriy: It's getting hard to follow which comment needs addressing. I'll refactor the `list-of` and `jeison-vector-of` processing. What else is outstanding?

BTW: Before you give me a hard time about using `apply`, `(vector '(some arbitrary list))` produces a vector of one element, vs. `(apply #'vector '(some arbitrary list))`, which produces a vector of three elements.

SavchenkoValeriy commented:
> @SavchenkoValeriy: It's getting hard to follow which comment needs addressing. I'll refactor the `list-of` and `jeison-vector-of` processing. What else is outstanding?
> 
> BTW: Before you give me a hard time about using `apply`, `(vector '(some arbitrary list))` produces a vector of one element, vs. `(apply #'vector '(some arbitrary list))`, which produces a vector of three elements.

Oh, yeah, I didn't check it or anything. `apply #'vector` is alright I guess, `vconcat` a bit better. Not a strong opinion though.

Anyways, let me just go through the hash table business a bit more thoroughly now.

SavchenkoValeriy reviewed (changes requested):
OK, I think that changes for `vector-of`, different json parsers and optional fields are ready to get merged (just replace `json-read-from-string` with `jeison--read-json-string` in tests beforehand so that they pass).

However, I can't accept hash table interface in its current form, it feels half-baked. We parse keys as strings, but in the hash-table they are suddenly symbols. It is disruptive and inconvenient for the user. The fact that the parser gives it to us this way is **not** an argument, we can convert it. 

We can do symbols instead of strings for keys and have a separate type constraint `key`. When we see it, we can check if it's a string and intern it. But the user should go back and forth between two types, it's not nice.

Also, I have a note about the documentation. I like the example now, it's way better than CMake 😅 But the code is far from being clear, and that's probably the last place we want to be unclear. Let's make it as concise and to the point as possible.

Sorry if I sound too harsh about this. I feel very passionate about clarity of APIs. So, please bear with me 🫣

bscottm commented:
Quick message since I'm on duty with the Navy this weekend, back on Monday.

Hash tables are not half baked. The JSON parsers convert the strings to symbols. If you RTSL, you'll notice that the code doesn't call `intern` anywhere -- it just slurps up the alist stream as presented. No magic, no special handling. Just placing the "keyword" and associated objects into the hash table.

Not a satisfying answer, I know. Not looking to modify the JSON parsers.

bscottm commented:
Output from `ielm`. Nota bene where strings have been converted to symbols.

```
ELISP> (jeison--read-json-string json-people)
((people .
  [((personal
     (address)
     (name
      (first . "John")
      (last . "Johnson")))
    (job
     (company . "Good Inc.")
     (title . "CEO")
     (location
      (country . "Norway")
      (city . "Oslo")))
    (skill-refs .
     ["skill-1" "skill-3"]))
   ((personal
     (address)
     (name
      (first . "Sammy")
      (last . "Smith")))
    (job
     (company . "Good Inc.")
     (title . "VP, Marketing")
     (location
      (country . "USA")
      (city . "Peoria")))
    (skill-refs .
     ["skill-2" "skill-3"]))])
 (skills
  (skill-1
   (name . "Programming")
   (description . "Can code like Richard S or Linus T"))
  (skill-2
   (name . "Design")
   (description . "Can think out loud, structure code systems"))
  (skill-3
   (name . "Communication")
   (description . "Can explain complicated things to social scientists"))))
```

SavchenkoValeriy commented:
> Quick message since I'm on duty with the Navy this weekend, back on Monday.
> 
> Hash tables are not half baked. The JSON parsers convert the strings to symbols. If you RTSL, you'll notice that the code doesn't call `intern` anywhere -- it just slurps up the alist stream as presented. No magic, no special handling. Just placing the "keyword" and associated objects into the hash table.
> 
> Not a satisfying answer, I know. Not looking to modify the JSON parsers.

It's not about modifying JSON parsers. Jeison wraps around JSON parsers, and we are free to do whatever we want with the parser output. Jeison *converts* parser output to be more user-friendly. It's the sole purpose of its existence. Hash table functionality in its current state is not user-friendly.

bscottm commented:
While true, you chose to use a lists, which both parsers use symbols to represent object keys.

SavchenkoValeriy commented:
And again, then intern the strings that intend to be keys. I want `oref` a field with a key and use it directly as a key. Only that way it is a consistent API. I don't care which type it is underneath, but it should be the same.

bscottm commented:
@SavchenkoValeriy: You're welcome to hack `symbol` into the `jeison--read-internal` `pcase` if you'd like to support a `list-of symbol`. However, adding `symbol` as a jeison type conversion makes the hash table example more complicated. The example will have to present both ways of accessing the hash table: by intern-ed string value and by symbol value. Then there's the problem of strings that are invalid symbols. Or the design choices you're now introducing -- are all strings in a list or object now symbols if they're potentially hash table keys when inside of hash table construction?

Just thinking this through a bit and looking gain better concept precision.

bscottm commented:
The alternative is applying `symbol-name` to the JSON member name to reconvert it back to a string. Or support both styles, as a string or as a symbol. A choice needs to be made, regardless.

(I do know the answer to my question. I just need to hear you express it. And I used to do this to Chandler Carruth when he was one of my interns.)

SavchenkoValeriy commented:
@bscottm I know I'm welcome to do changes, it's my repo after all. 

Let's make a step back. You propose a change and I disagree on the interface of the new feature. I want the example in the README section to look like this:
```emacs-lisp
(setq people-data (jeison-read jeison-people json-people))
(setq people (oref people-data people))
(setq skills (oref people-data skills))
(mapcar
 (lambda (person)
   (format "%s skills: %s\n"
           (oref person name)
           (mapconcat (lambda (skill-ref)
                        (oref (gethash skill-ref skills) name))
                      (oref person skill-refs)
                      ", ")))
 people) ;; => ("Johnson skills: Programming, Communication" "Smith skills: Design, Communication")
 ```
 
This is concise and consistent. All the rest is implementation detail. If you convert the symbols to strings or strings to symbols, it doesn't matter, as long as it is well tested and provides the desired interface.

If you ask me, I would've probably solved this problem without hash tables to begin with. I would've gone with providing information about how to retrieve each skill and *inlined* the data removing this indirection. It has a lot of pitfalls and questions about how to make it concise and consistent. But it's a different question. Here we have your change that you want to merge, and I'm telling you that it's not there yet.

bscottm commented:
@SavchenkoValeriy: What you've expressed is a desire, not a spec for what you think is a solution. What I needed was a concise spec -- after all, you want elegant APIs, so giving me a proposed API spec would have been helpful and not a waste of my time coding something that you'll end up nitpicking, which undoubtedly, you'll do with the next patch (which I've just pushed.)

SavchenkoValeriy commented:
@bscottm thanks. It looks like we finally got there functionality-wise. And yes, I need to go through the code one more time. It's called *code review*. After a fast glance, I can already see that there is still no test for failing `hash-table` type check and some minor inconsistencies with examples in the README. It's getting there, it is converging.

NOTE: Maybe it would've been faster not to have 5 features in one PR, or actually fixing my comments the first time I bring them up.

Cheers!

bscottm commented:
Just give me a list (not separate code review comments) to work off. You might want to approve the workflow so I can correct deficiencies that might arise there.

SavchenkoValeriy commented on README.md:
```diff
@@ -1,6 +1,65 @@
+    a slot to `nil`. `jeison-read` can parse and subsequently set the slot to
+    `type`.
+
+    The primary use for `:type (or null type)` is handling optional
```
I believe that a separate section further in the document is sufficient, and we can keep **New in 1.1.0** short and cut everything starting from here.

SavchenkoValeriy commented on README.md:
````diff
@@ -295,16 +398,380 @@ Here is an example:
+symbols as keys:
+
+```emacs-lisp
+(let* ((people-data (jeison-read jeison-people json-people)))
````
Can you please simplify this code to be less nested and to match the style of other examples? 
I suggest to add `setq`s for `people-data`, `people` and `skills`, and remove explicit prints.

SavchenkoValeriy commented on README.md:
````diff
@@ -295,16 +398,380 @@ Here is an example:
+
+```
+Johnson skills: Programming, Communication
+Smith skills: Design, Communication
````
I suggest not to print the hash table and only keep these two lines. And it should also match the output style of the rest of the document, ie to be inline with the code via `;; =>` .

SavchenkoValeriy commented on README.md:
````diff
@@ -295,16 +398,380 @@ Here is an example:
+```
+
+The contents of the `skills` hash table are pretty-printed to show the
+keys are symbols.
````
Without hash table printout, this paragraph won't be needed.

SavchenkoValeriy commented on README.md:
```diff
@@ -295,16 +398,380 @@ Here is an example:
+
+`jeison-hash-table-of` also accepts `string` for the hash table's key
+type. Simply change the definitions of `jeison-people` and
+`jeison-person` to use `string` instead of `symbol`:
```
This and the examples below don't add a lot of value, but bloat the documentation making hash tables look more complicated than they actually are. I suggest cutting all and keep only the information that we could've also used `string` type there. Maybe even rephrase it and explicitly say that it accepts ONLY strings and symbols as keys.

SavchenkoValeriy replied:
I still believe that having two almost identical examples is unnecessary for understanding. It bloats the documentation and makes this feature seem more complex than it is.

SavchenkoValeriy commented on README.md:
```diff
@@ -295,16 +398,380 @@ Here is an example:
+is a slight performance advantage to using symbols because the hash
+table uses `eq` for comparisons, vice `equal` for strings. However, if
+the hash table has a short lifetime, the performance advantage is
+likely minimal.
```
I like this paragraph, I only suggest to mark it as a note.

SavchenkoValeriy commented on README.md:
````diff
@@ -295,16 +398,380 @@ Here is an example:
+Johnson skills: Programming, Communication
+Smith skills: Design, Communication
+Twiddler skills: Missing or unknown skills
+```
````
This is a really huge example for a very simple and straightforward feature. I also expect users not to read this file end-to-end (hash tables is quite an unusual use case), so adding dependency on a less essential feature is not helping. I suggest to decouple them and drastically cut the example here.

SavchenkoValeriy commented on jeison.el:
```diff
@@ -26,14 +28,86 @@
-(require 'json)
+
+(defvar jeison-debug-mode nil
+  "'jeison' debugging level.  0 = off, 5 = highest.")
```
Seems to be unused

bscottm replied:
It doesn't hurt to keep the `jeison-debug-mode` if it may be needed at a later time. Suggest keeping it.

SavchenkoValeriy replied:
OK

SavchenkoValeriy commented on test/jeison-test.el:
```diff
@@ -218,4 +234,109 @@
+                     (not (null (gethash 'codemodel-v2 client-1)))
+                     (not (null client-2))
+                     (null (gethash 'non-existent client-table))))))))
+
```
It is possible to specify hash table type for something that doesn't match our expectations. Failed type checks should also get tested for.

bscottm replied:
Test meets my expectations. What did you really mean? (I'm mentoring you here: be precise and concise.)

SavchenkoValeriy replied:
Judging by the number of comments in this PR, the mentoring goes the other way.

Please add tests for situations when `jeison-hash-table-of` type check throws an error:

1. key type mismatch
2. value type mismatch
3. unsupported key-type, eg `jeison-hash-table-of number string`

bscottm replied:
I think you have a misunderstanding with respect to type checking, CLOS (EIEIO) and Lisp/Elisp hash tables. You can `puthash` almost anything as a key. I can `oref` the hash table, directly `puthash` strings as keys and `gethash` will succeed, even though `jeison-read` read a class that has a hash table that uses symbols. Lisp/Elisp hash tables can have a mix of symbols and strings (and integers and classes and...) I could call `cl-check-type` directly, but that defeats the point of the test. EIEIO calls `cl-type-check` infrequently on an instance's contents.

Key and value mismatches are pointless. An unsupported key type -- you'd only indirectly alluded to key and value mismatches. Sure, I'll construct a pedantic test for an invalid key type, which tests EIEIO's type constraint system when instances are created. `jeison` doesn't emit the `signal` or `error`; EIEIO emits it.

SavchenkoValeriy replied:
Or you misunderstand static typing annotations on dynamic languages and guarantees that Jeison provides.

About hash tables, we can use **exactly** the same argument for lists. List is a heterogenous collection, but at the same time `list-of TYPE` is a useful check. Why? It provides additional guarantees about what you might find in such list. Jeison uses it to check user expectations about the JSON they are trying to parse. That's it, early error to save user a headache.

Now let's go to user expectations and `jeison-hash-table-of number string`. I say that this user misunderstood how hash table parsing from JSON works and the API needs to surface an error about it as early as possible. The answer "Something will go wrong later down the line" is bad and unhelpful.

Let's move to another example:
```emacs-lisp
(jeison-defclass jeison:jeison-class nil ((x :initarg :x :type (jeison-hash-table-of string string)))
(jeison-read jeison:jeison-class "{\"x\": 42}")
```
If the user constructed this Jeison class and later receives this JSON file, we **have to** report it. Reality doesn't match the expectation. It doesn't matter if we check it in our code or if EIEIO checks it for us. It doesn't matter that the user can manually `oref` that hash table later, and `puthash` whatever they want. Type annotations is the feature of Jeison and we need to test it.

You implemented `jeison-hash-table-of` constraint, so it's not the test of EIEIO constraint system, it is a test for your code. I want to remind you that it DID HAVE A BUG. And right now there is no test covering that logic.

SavchenkoValeriy reviewed (commented):
This is the last chunk of my comments. I'm happy with the implementation itself, great job!
I went through the README and added a few suggestions on where it should be changed (mostly editing it down). Also an important test case missing and should be added.

Once these are fixed, I'd be happy to accept. Thanks!

bscottm committed:
doc updates...

bscottm committed:
Add invalid key-type test

- jeison:check-hash-table-table-of-invalid-key-type tests an invalid
  key type (integer) as the key-type argument to jeison-hash-table-of.

  NOTE: EIEIO infrequently checks the instance's type constraint; this
  definitely occurs during make-instance, but not observed
  elsewhere. Comment in code annotates when EIEIO signals
  invalid-slot-type.

- Make the jeison-hash-table-of type constraint code more compact and
  efficient.

- Reduce jeison's cl-lib overall overhead with eval-when-compile. It
  only uses cl-lib's macros, so no reason to incur needless overhead.

SavchenkoValeriy commented on test/jeison-test.el:
```diff
@@ -234,6 +234,32 @@
+                   (progn
+                     (jeison-defclass jeison:treif-key-type-class ()
+                       ((foo :initarg :foo :type (jeison-hash-table-of integer string))))
+                     (make-instance jeison:treif-key-type-class :foo (make-hash-table))
```
I see why you hesitated, it is indeed weird to check `make-instance` since it's not part of Jeison.
Jeison checks (or should check) for expected types in `jeison-read` and that's what should be tested.

bscottm replied:
`make-instance` is the *only* way to trigger the `invalid-slot-type` signal and EIEIO type checking, which is what I took away from your comments. Now do you understand why I asked you to be precise?

You want (a) code in `jeison--read-internal` to validate the key type is either `string` or `symbol`, (b)  a test to exorcise that code (yes, the pun is intended.)

SavchenkoValeriy replied:
It is again a discussion similar to hash table keys. No need to go down to details, look at it from the interface perspective. The user has expected types and some JSON. Jeison maps JSON to user expectations. If the expectation is not correct, it is Jeison's job to surface that.

Let's look at this code here: https://github.com/SavchenkoValeriy/jeison/blob/19a51770f24eaa7b538c7be6a8a5c25d154b641f/jeison.el#L170-L173
That is one example of where we do it.

It is possible that the user expectations expressed in the form of static types do not match the actual JSON. Jeison does have tests for this, here's an example:
https://github.com/SavchenkoValeriy/jeison/blob/19a51770f24eaa7b538c7be6a8a5c25d154b641f/test/jeison-test.el#L114-L120

With new types, user expectations can also be wrong. It is APIs job to surface any problem we can spot. Let me repost that snippet from before:
```emacs-lisp
(jeison-defclass jeison:jeison-class nil ((x :initarg :x :type (jeison-hash-table-of string string)))
(jeison-read jeison:jeison-class "{\"x\": 42}")
```
`42` cannot be parsed as a hash table at all, we need to report it.
Let's go further:
```js
{"x": [0, 1], "y": [2], "z": [3, 4, 5]}
```
It looks like a hash table, now! But if the user asked for `jeison-hash-table-of string string`, we have a problem and we need to report it.
And yet another example:
```js
{"x": "a", "y": [2], "z": "b"}
```
that we can use to test type mismatch for one of the values.

When the user writes `jeison-hash-table-of NOT_STRING_OR_SYMBOL T`, it is another example of wrong expectations. Yes, it is different from other cases. Yes, we can check it not during the reading part, but during the class declaration. But it's all implementation details. The main part is that it's wrong, it should be checked, it should be reported, it should be tested.

I want tests for that, and not for `make-instance` or `gethash`.

bscottm committed:
Extra pedantry

Add test code to show that 'puthash' isn't picky about what it'll
stash in a hash table and that EIEIO doesn't actually check the type
constraint.

SavchenkoValeriy commented on test/jeison-test.el:
```diff
@@ -218,4 +234,148 @@
+                     (null (gethash "client-123456" client-table))
+                     (null (gethash "client-7890ab" client-table))
+                     ;; And neither do vectors.
+                     (null (gethash [1 3 5] client-table))))))))
```
Now this is pure trolling. Try listen to other people instead of trying to mentor them (I'm mentoring you here).
Jeison reading functions DO TYPE CHECKING, you introduced NEW TYPES, test that the type checking for new types is WORKING when we read from JSON.

bscottm committed:
More specific jeison--read-internal hash table testing

- Tighten jeison--read-internal's hash table key-type check; don't
  leave it to EIEIO's cl-check-type during make-instance.

- Exercise the jeison-wrong-parsed-type signal with hash table values.

SavchenkoValeriy commented on test/jeison-test.el:
```diff
@@ -234,11 +234,14 @@
                           '(a (jeison:filter-candidates b) number))))))
 
-(ert-deftest jeison:check-hash-table-of-invalid-key-type ()
+(ert-deftest jeison:check-cl-deftype-jeison-hash-table-of ()
```
This test is incorrect, it doesn't test Jeison's type checking functionality. Refer to my previous comments for explanation.

SavchenkoValeriy commented on test/jeison-test.el:
```diff
@@ -326,20 +317,43 @@
+    ((a :initarg :a :initform "default" :type string)
+     (b :initarg :b :initform 'undefined :type symbol)
+     (c :initarg :c :initform [] :type (jeison-vector-of integer))))
+  (let ((json-stream "{ \"key-1\": { \"a\" : \"a's value\", \"b\": \"b-symbol-value\", \"c\": [99, 100, 101] } }"))
```
Good, this test matches one of the tests that I asked for. However, the input here still looks like a hash-table, so it checks if the values of something that looks like a hash-table match user expectations. I asked for another test, where the value in the specified path DOES NOT LOOK like hash-table, ie is not parsed as an `alist`.

bscottm commented:
@SavchenkoValeriy: Merge or close the pull request. My patience wears thin with your condescension and pedantry.

bscottm committed:
Proper polymorphic dispatch

SavchenkoValeriy commented:
@bscottm excuse me, but I won't let you to bully me into merging this PR. Fix the comments, and I'll merge.

> My patience wears thin with your condescension and pedantry.

Likewise.

bscottm commented:
Godspeed.
